### PR TITLE
[CBRD-23687] Apply SSL Broker/CAS to damson

### PR DESCRIPTION
**Pull request by kisoo-han**
http://jira.cubrid.org/browse/CBRD-23687
* Apply Secure Broker/CAS (SSL) into develop

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/broker/cas_ssl.c (1 incorrect indentation(s) found)
```
   174       cas_log_write_and_end (0, true, "SSL: Accept failed - '%s'", ERR_error_string (err, NULL));
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/broker/cas_ssl.c [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+#define SOCKET_NONBLOCK		1
+#define SOCKET_BLOCK		0
+
+static SSL *ssl = NULL;
+bool ssl_client = false;
+
+static int cas_ssl_validity_check (SSL_CTX * ctx);
+
+int
+initSSL (int sd)
```
[the remark this answers is not in the document]
Comment (kisoo-han) [reply src/broker/cas_ssl.c]: Naming was changed to cas_init_ssl (), it looks better. By the way, I will seriously examine and testing **cas_network.c** and **cas_ssl.c** including error handling, resource release as well as removing unnecessary codes inside it.
Thank you for your detail review!

File: src/broker/cas_ssl.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+  if ((ssl = SSL_new (ctx)) == NULL)
+    {
+      cas_log_write_and_end (0, true, "SSL: Creating SSL context failed.");
+      return ER_SSL_GENERAL;
+    }
+
+  if (SSL_set_fd (ssl, sd) == 0)
+    {
+      cas_log_write_and_end (0, true, "SSL: Cannot associate with socket.");
```
[the remark this answers is not in the document]
Comment (kisoo-han) [reply src/broker/cas_ssl.c:163]: Changed by your comment.

File: src/broker/cas.c [hunk 3]
```diff
@@ -124,6 +124,10 @@ static void set_db_connection_info (void);
 static void clear_db_connection_info (void);
 static bool need_database_reconnect (void);
 
+extern int cas_init_ssl (int);
```
Comment (mhoh3963) [inline src/broker/cas.c:127]: how about change order of definition ?
extern bool ssl_client;
extern int cas_init_ssl (int);
extern void cas_ssl_close (int client_sock_fd);

Comment (kisoo-han) [reply src/broker/cas.c:127]: ok! I will accept it.

File: src/broker/cas_ssl.c [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      return ER_SSL_GENERAL;
+    }
+
+  return 0;
+}
+
+bool
+is_ssl_data_ready (int sock_fd)
+{
+  return (SSL_has_pending (ssl) == 1 ? true : false);
```
Comment (mhoh3963) [inline src/broker/cas_ssl.c:294]: SSL_has_pending() returns the number of bytes which have been processed, buffered, so I think you should change this code as the following. 

return (SSL_has_pending (ssl) >= 1 ? true : false)

File: src/broker/cas_network.c [hunk 5]
```diff
@@ -581,13 +605,13 @@ read_buffer (SOCKET sock_fd, char *buf, int size)
...
 	  /* RECEIVE NEW REQUEST */
-	  read_len = READ_FROM_SOCKET (sock_fd, buf, size);
+	  read_len = READ_FROM_NET (sock_fd, buf, size);
 #if defined(ASYNC_MODE)
 	}
     }
 #endif /* ASYNC_MODE */
 
-  if (read_len <= 0)
+  if ((!ssl_client && read_len <= 0) || (ssl_client && read_len < 0))
```
Comment (mhoh3963) [inline src/broker/cas_network.c:614]: When ssl_client is true, Is that read_len is 0 no problem ?
SSL_read() success if return value great than 0 (> 0).

File: src/broker/cas_ssl.c (hunk 4 above)
Comment (kisoo-han) [reply src/broker/cas_ssl.c:294]: SSL_has_pending () returns value 0 or 1.
I think you talking about SSL_pending ()
either 

```
return (SSL_has_pending (ssl) == 1 ? true : false);
or
return (SSL_pending (ssl) >= 1 ? true : false);
```

Comment (mhoh3963) [reply src/broker/cas_ssl.c:294]: I got it.

File: src/broker/cas_network.c (hunk 5 above)
Comment (kisoo-han) [reply src/broker/cas_network.c]: I misunderstood the SSL_read manual. 
You are right. SSL_read returns 0 means the connection was closed, and error.
I will fix it. Thank you for your comment.

Comment (mhoh3963) [reply src/broker/cas_network.c]: It's my pleasure.

**Pull request merged**